fn_utilities/fn_utilities/components/utilities_attachment_to_base64.py

```diff
@@ -23,35 +23,55 @@ def _attachment_to_base64_function(self, event, *args, **kwargs):
             incident_id = kwargs.get("incident_id")  # number
             task_id = kwargs.get("task_id")  # number
             attachment_id = kwargs.get("attachment_id")  # number
+            artifact_id = kwargs.get("artifact_id") # number
 
             log.info("incident_id: %s", incident_id)
             log.info("task_id: %s", task_id)
             log.info("attachment_id: %s", attachment_id)
-            if incident_id is None and task_id is None:
-                raise FunctionError("Error: incident_id or task_id must be specified.")
-            if attachment_id is None:
-                raise FunctionError("Error: attachment_id must be specified.")
+            log.info("artifact_id: %s", artifact_id)
+            
+            input_err_msg = lambda msg: "Error: {0} must be specified.".format(msg)
+
+            if incident_id is None:
+                raise FunctionError(input_err_msg("incident_id"))
+            elif attachment_id is None and artifact_id is None:
+                raise FunctionError(input_err_msg("attachment_id or artifact_id"))
 
             yield StatusMessage("Reading attachment...")
+            
+            client = self.rest_client()
+            metadata = None
+
             if task_id:
                 metadata_uri = "/tasks/{}/attachments/{}".format(task_id, attachment_id)
                 data_uri = "/tasks/{}/attachments/{}/contents".format(task_id, attachment_id)
+            elif artifact_id:
+                metadata_uri = "/incidents/{0}/artifacts/{1}".format(incident_id, artifact_id)
+                artifact = client.get(metadata_uri)
+                metadata = artifact.get('attachment') # only interested in attachment of the artifact
+                if metadata:
+                    data_uri = "/incidents/{0}/artifacts/{1}/contents".format(incident_id, artifact_id)
+                else:
+                    raise FunctionError("Artifact has no attachment or supported URI")
             else:
                 metadata_uri = "/incidents/{}/attachments/{}".format(incident_id, attachment_id)
                 data_uri = "/incidents/{}/attachments/{}/contents".format(incident_id, attachment_id)
 
-            client = self.rest_client()
-            metadata = client.get(metadata_uri)
+            # Condition prevents a duplicate call
+            if metadata is None:
+                metadata = client.get(metadata_uri)
             data = client.get_content(data_uri)
 
             results = {
                 "filename": metadata["name"],
                 "content_type": metadata["content_type"],
                 "size": metadata["size"],
                 "created": metadata["created"],
-                "content": base64.b64encode(data)
+                "content": str(base64.b64encode(data))
```

Comment:
Why do we need to add in this `str()` casting?

Reply:
Without casting Base64 object to string if we send results json then it throws a serialization error.

Reply:
OK. Has it been tested in Python2.7 and Python3.6? I just wonder why we did not have to cast it before this.

Also make sure that we test this with attachments that have text that contain unicode. You could create a text file containing some characters from this site: https://www.ltg.ed.ac.uk/~richard/unicode-sample-3-2.html and ensure the function works on it

Reply:
ok